Workspace support

UebelAndre opened the pull request:
The following pull requests are included in this change set and should be closed first:
- https://github.com/google/cargo-raze/pull/275
- https://github.com/google/cargo-raze/pull/268

This pull requests adds support for [cargo workspaces](https://doc.rust-lang.org/book/ch14-03-cargo-workspaces.html) and closes https://github.com/google/cargo-raze/issues/262

### Notable Changes:
1. There is now a new setting `workspace_member_dir` which is used to determine the location where a package's dependencies will get aliased. Previously, a BUILD file would be rendered into `workspace_path` which would contain aliases for direct dependencies. Because we now account for there being multiple packages (ie, in a workspace) a directory matching the name specified by `workspace_member_dir` will be created in each package directory (where a `Cargo.toml` file lives) containing a BUILD file with the aliases. This is intended to make it easier for users to specify the dependencies they care about as well as maintain the previous behavior (`workspace_member_dir` should simply match the package name specified by `workspace_path` to result in no change for existing projects).
2. Binary dependencies are treated as workspace members. This is done by downloading the binary dependency's source into the mock workspace when `raze` is gathering metadata and modifying the workspace root's `Cargo.toml` to include a [workspace](https://doc.rust-lang.org/cargo/reference/workspaces.html) definition that includes the binary dependency. This allows `raze` to share a single lockfile for all dependencies and ensure all dependencies use compatible versions with each other. (this actually solves a bug I noticed when generating `wasm-bindgen-cli` for `rules_rust`. [This dependency](https://github.com/bazelbuild/rules_rust/blob/master/wasm_bindgen/raze/Cargo.lock#L37-L44) will update if the file is regenerated and cause a build error). `Cargo-raze` will now output a `Cargo.raze.lock` file when binary dependencies are specified. This file is authoritative over subsequent raze runs.
3. Add support for defining Raze Settings via `[workspace.metadata.raze]`. This will take precedence over package definitions (`[package.metadata.raze]`) if it's defined. This should provide a consistent central location for settings even in projects that only provide a single root package.


Sorry about the large PR. I did a pass to split out what I could into other pull requests but I found it difficult to do so.

UebelAndre commented:
@acmcarther sorry about the large PR. Looking forward to your feedback. I'm really hoping you like some of the changes here 😄

UebelAndre commented:
@wchargin do you have any suggestion on how `[package.metadata.raze]` would work in the context of a workspace? I'm getting warnings output in #277

UebelAndre commented:
> @wchargin do you have any suggestion on how `[package.metadata.raze]` would work in the context of a workspace? I'm getting warnings output in #277

Ah, found https://doc.rust-lang.org/cargo/reference/workspaces.html#the-workspacemetadata-table

Sorry to bother 😅

acmcarther commented on impl/src/metadata.rs:
```diff
@@ -50,25 +97,132 @@ pub struct CargoWorkspaceFiles {
+    let re = Regex::new(r".+\(path\+file://(.+)\)")?;
+    for member in no_deps_metadata.workspace_members.iter() {
+      // Get a path to the workspace member directory
+      let path = {
```
More descriptive name?

acmcarther commented on impl/src/metadata.rs:
```diff
@@ -50,25 +97,132 @@ pub struct CargoWorkspaceFiles {
+      let toml_path = path.join("Cargo.toml");
+
+      // If no Cargo.toml file is found, skip this
+      if !toml_path.exists() {
```
What does it mean if there's no toml path? Should we log? Is it user error? Is it sometimes expected?

acmcarther replied:
Ignored this comment?

UebelAndre replied:
Added documentation and updated this to an error

acmcarther commented on impl/src/metadata.rs:
```diff
@@ -50,25 +97,132 @@ pub struct CargoWorkspaceFiles {
+
+              self.make_symlink(&path, &new_path)?;
+            },
+            _ => {},
```
What situations are caught here? when does this entry not contain Ok?

UebelAndre replied:
https://docs.rs/glob/0.3.0/glob/ suggests there would be errors in some cases. I opted to ignore them but could escalate that error. Your thoughts?

UebelAndre replied:
Updated implementation

acmcarther commented on impl/src/metadata.rs:
```diff
@@ -50,25 +97,132 @@ pub struct CargoWorkspaceFiles {
+
+  // Create a symlink file on windows systems
+  #[cfg(target_family = "windows")]
+  fn make_symlink(&self, src: &Path, dest: &Path) -> Result<()> {
```
Thanks for handling this for all platforms!

acmcarther commented on impl/src/metadata.rs:
```diff
@@ -50,25 +97,132 @@ pub struct CargoWorkspaceFiles {
+      let mut r_url = self.registry_url.clone();
+      r_url.set_path("");
+      r_url.to_string()
+    };
```
Comment above this block? I found the `set_path("")` suspicious.

acmcarther commented on impl/src/metadata.rs:
```diff
@@ -92,189 +246,203 @@ impl CargoMetadataFetcher {
   }
 
-  pub fn fetch_crate_src(
+  fn inject_binaries_into_workspace(
```
I think this needs a comment -- the name is provocative but not really self explanatory.

acmcarther commented on impl/src/metadata.rs:
```diff
@@ -92,189 +246,203 @@ impl CargoMetadataFetcher {
+
+  fn fetch_crate_checksum(&self, name: &str, version: &str) -> Result<String> {
+    let index_path_is_url = url::Url::parse(&self.index_url).is_ok();
+    let crate_index_path = if index_path_is_url {
```
I think this is the second time this review I"m seeing this code block.

Two things.

1. Should this be moved into a function?
2. Should we just be opening this index in `new` or something and holding onto it?

UebelAndre replied:
Where is the second instance of this code block?

acmcarther reviewed (changes requested):
I got tired half way through the review  (on me, not you, didn't get enough sleep last night), so I'm only half way done. I'll need to continue from about half way through impl/src/metadata.rs on my next review.

Looks good so far. Thank you for your continued work here and throughout the repo!

UebelAndre commented on impl/Cargo.toml:
```diff
@@ -34,6 +34,8 @@ docopt = "1.1.0"
 glob = "0.3.0"
 itertools = "0.9.0"
 log = "0.4.11"
+pathdiff = "0.2.0"
```
There's very likely an smarter implementation that I didn't think of that avoids this dependency all together. I have the same concern but felt for the time being, it'd be better to use a tested crate than hand write something.

UebelAndre commented on impl/src/bin/cargo-raze.rs:
```diff
@@ -65,20 +66,26 @@ Options:
 "#;
 
 fn main() -> Result<()> {
+  // Parse options ---
```
Was just trying to highlight the "sections" within this tool. Any suggestions? (I can still remove the `---`)

UebelAndre commented:
@acmcarther Hey, I've addressed some stuff here and had questions on a few things but is otherwise ready for another pass 😀

acmcarther commented on impl/src/rendering/bazel.rs:
```diff
@@ -220,18 +219,37 @@ impl BuildRenderer for BazelRenderer {
+      let all_packages: Vec<CrateContext> = planned_build
+        .crate_contexts
+        .iter()
+        .filter_map(|ctx| {
```
I think this can be expressed with filter:  ctx.is_binary_dependency || ctx.workspace_member_dependents.contains(member_path)

acmcarther commented on impl/src/rendering/bazel.rs:
```diff
@@ -275,18 +294,36 @@ impl BuildRenderer for BazelRenderer {
-        message: unwind_tera_error!(e),
-      })?;
+    for member_path in planned_build.workspace_context.workspace_members.iter() {
+      let all_packages: Vec<CrateContext> = planned_build
```
This block seems to duplicate the code above.

UebelAndre replied:
Can you elaborate?

acmcarther replied:
L222
```
    for member_path in planned_build.workspace_context.workspace_members.iter() {
      let all_packages: Vec<CrateContext> = planned_build
        .crate_contexts
        .iter()
        .filter(|ctx| {
          ctx.is_binary_dependency || ctx.workspace_member_dependents.contains(member_path)
        })
        .map(|ctx| ctx.clone())
        .collect();
      let rendered_alias_build_file = self
        .render_aliases(&workspace_context, &all_packages)
        .map_err(|e| RazeError::Rendering {
          crate_name_opt: None,
          message: unwind_tera_error!(e),
        })?;

      file_outputs.push(FileOutputs {
        path: render_details
          .cargo_root
          .join(member_path)
          .join(&render_details.workspace_member_output_dir)
          .join("BUILD.bazel"),
        contents: rendered_alias_build_file,
      });
    }
```

L292
```
      let all_packages: Vec<CrateContext> = planned_build
        .crate_contexts
        .iter()
        .filter_map(|ctx| {
          if ctx.is_binary_dependency {
            Some(ctx.clone())
          } else if ctx.workspace_member_dependents.contains(member_path) {
            Some(ctx.clone())
          } else {
            None
          }
        })
        .collect();
      let rendered_alias_build_file = self
        .render_remote_aliases(&workspace_context, &all_packages)
        .map_err(|e| RazeError::Rendering {
          crate_name_opt: None,
          message: unwind_tera_error!(e),
        })?;

      file_outputs.push(FileOutputs {
        path: render_details
          .cargo_root
          .join(member_path)
          .join(&render_details.workspace_member_output_dir)
          .join("BUILD.bazel"),
        contents: rendered_alias_build_file,
      });
```

acmcarther commented on impl/src/planning/subplanners.rs:
```diff
@@ -232,70 +231,19 @@ impl<'planner> WorkspaceSubplanner<'planner> {
+        .iter()
+        .sorted_by_key(|n| &n.id)
+        .filter_map(|node| self.create_crate_context(node, &self.crate_catalog))
+        .collect::<Result<Vec<CrateContext>>>()?,
```
I'd prefer more imperative framing of this. I don't think the fluent writing of this is is very readable.

UebelAndre replied:
This is only a slight modification to what was there before. Can this be deferred or can you write a suggestion?

GregBowyer replied:
Minor: `Ok(` wrap on something that returns `Result` ? Why not unfold one level?

UebelAndre replied:
Done

acmcarther commented on impl/src/metadata.rs:
```diff
@@ -50,25 +97,132 @@ pub struct CargoWorkspaceFiles {
+        let re_match = match capture.get(1) {
+          Some(re_match) => re_match,
+          None => continue,
+        };
```
Why not my version?

acmcarther replied:
There must be something wrong with the UI. Try opening the diff view. Do you see the full comment chain then?

Reproducing the comment:

This above can optionally be condensed.
```
let filename = {
  let filename_match =  re.captures(&member.repr).and_then(|c| c.get(1));
  if (filename_match.is_none()) {
    continue;
  }
  PathBuf::from(filename_match.unwrap()); // UNWRAP: Guarded above
};
```
(i don't really feel strongly about the forumulation of the unwrap -- more important is that you can use a map to avoid the second match)

acmcarther commented on impl/src/bin/cargo-raze.rs:
```diff
@@ -89,77 +96,81 @@ fn main() -> Result<()> {
     lock_path_opt,
   };
 
+  let baze_root = find_bazel_workspace_root().unwrap_or(std::env::current_dir()?);
```
why was this marked resolved.

UebelAndre replied:
I changed it but must have stomped the changes out when rebasing or some thing. I'll update this again

UebelAndre replied:
I don't know why it doesn't show up in this diff but I added a comment above it explaining what the variable is. Is that not sufficient?

acmcarther replied:
Comment was regarding name of variable.

UebelAndre replied:
Updated the variable name.

acmcarther reviewed (commented):
This is going to take a very long time to review.

UebelAndre commented:
> This is going to take a very long time to review.

Why is that? Is there something I can do to help with that?

UebelAndre commented on impl/src/settings.rs:
```diff
@@ -384,17 +402,38 @@ pub fn load_settings<T: AsRef<Path>>(cargo_toml_path: T) -> Result<RazeSettings,
+        }),
+      raze: top_level_raze,
+      ..
+    }) => {
```
The logic for loading settings is something I'm hoping to improve in a future PR. I want to make update this to be smarter by parsing `workspace.metadata.raze` and `package.metadata.raze` and combining the results. Can we please defer this to another PR?

GregBowyer replied:
Yeah I can be the king of crazy patterns and even for me this is a bit much.

UebelAndre replied:
What can I do do simplify this without having to write a large amount of new logic?

UebelAndre replied:
Suggestions are greatly welcomed 😅 🙏

UebelAndre replied:
I've made a change but have no idea if it's better or not 😞

UebelAndre commented:
https://github.com/google/cargo-raze/pull/276#discussion_r520806557
@acmcarther what are you referring to here?

UebelAndre commented on impl/src/testing.rs:
```diff
@@ -102,13 +142,21 @@ pub fn make_basic_workspace() -> (TempDir, CargoWorkspaceFiles) {
+  name: &str,
+  version: &str,
+  mock_server: &'server MockServer,
+) -> (TempDir, Vec<MockRef<'server>>) {
```
This is a pattern that existed in the project before I started working on this project: https://github.com/google/cargo-raze/blob/master/impl/src/testing.rs#L72-L99

Could this not be considered acceptable in test code? I'm happy to update documentation for this function to explain the return type.

acmcarther replied:
Hmm.

That is an entirely valid and powerful rebuttal, but I think that case is slightly different because it returns two self-descriptive types. These are not quite self descriptive.

UebelAndre replied:
Updated

UebelAndre commented:
> > This is going to take a very long time to review.
> 
> Why is that? Is there something I can do to help with that?

Oh, is it just that you're busy @acmcarther ? Totally understandable if so.

Also, this PR is ready for another review

acmcarther commented on impl/src/bin/cargo-raze.rs:
```diff
@@ -89,77 +96,81 @@ fn main() -> Result<()> {
+  }
+
+  for output in bazel_file_outputs.iter() {
+    if options.flag_dryrun.unwrap_or(false) {
```
Ignored this comment?

UebelAndre replied:
Sorry, I don't know why this disappeared from the overview.

acmcarther commented on impl/src/metadata.rs:
```diff
@@ -19,16 +19,67 @@ use std::{
+use url::Url;
 
-use rustc_serialize::hex::ToHex;
+pub use cargo_metadata::{DependencyKind, Metadata, Node, Package, PackageId};
```
Why pub use?

acmcarther replied:
(is this because you wanted to access it from other modules through this module? if so, I'd really prefer you just import it like normal. If it's because you want the library to export this type, I also find that questionable -- I don't think it's a great idea to export other libraries types through our interface)

UebelAndre replied:
I was thinking of this from the perspective of external crates (things that might import `cargo-raze`). The `RazeMetadata` struct is the source of truth so I thought whatever it was using should be re exported. I thought this was a common pattern since I'd run into it in the `cargo_lock` and `cargo_metadata` crates in how they handle `semver::Version`. Should I still change this?

GregBowyer replied:
I would think its best for now to not be pub, I dont feel there is a good surface API in cargo-raze for exporting.

UebelAndre replied:
I think that's worth exploring but I think that should be a focused effort. I've made these not public.

UebelAndre replied:
Oh, and thanks @GregBowyer for checking out this PR!

I'm particularly interested in hearing feedback on the functionality in here. Would love to get your thoughts/review 😄

acmcarther commented on impl/src/metadata.rs:
```diff
@@ -19,16 +19,58 @@ use std::{
+}
+
+impl RazeMetadata {
+  pub fn new(
```
Ignored this comment?

acmcarther replied:
I don't feel strongly about this, but I'm really surprised you're not seeing these comments. I just click the `Files Changed` tab. Perhaps you're not expanding the unexpanded files?

UebelAndre replied:
`checksums` was not pub but I can make it so. I thought keeping it private would drive users to use `get_checksum` instead of directly accessing the field but I suppose it's not that valuable.

acmcarther commented on impl/src/metadata.rs:
```diff
@@ -19,16 +19,58 @@ use std::{
+    }
+  }
+
+  pub fn get_checksum(&self, name: &str, version: &str) -> Option<&String> {
```
Ignored this comment?

acmcarther commented on impl/src/metadata.rs:
```diff
@@ -19,16 +19,58 @@ use std::{
+  }
+}
+
+/** A struct containing information about a binary dependency */
```
Ignored this comment?

UebelAndre replied:
Moved it back (I think to the right spot)

acmcarther commented on impl/src/metadata.rs:
```diff
@@ -37,7 +79,12 @@ const SYSTEM_CARGO_BIN_PATH: &str = "cargo";
+    &self,
+    files: &CargoWorkspaceFiles,
+    binary_dep_info: Option<&HashMap<String, cargo_toml::Dependency>>,
+    override_lockfile: Option<PathBuf>,
```
Ignored this comment?

UebelAndre replied:
I don't think that `override_lockfile` should go with `CargoWorkspaceFiles` since it's intended to be a unique lockfile only generated by `cargo-raze`. The reason this file exists in the first place was `(b)`. That information does not exist within the projects cargo metadata because of the inconsistency in how Cargo handles binary dependencies. Raze generates a new lockfile to make sure subsequent runs are reproducible.

I actually feel that the `CargoWorkspaceFiles` struct should no longer exist. https://github.com/google/cargo-raze/pull/274 introduced a way to cleanly embed Raze settings into cargo metadata and because of this I feel we should actually be using a [`MetadataCommand`](https://docs.rs/cargo_metadata/0.12.0/cargo_metadata/struct.MetadataCommand.html) to load settings. It should no longer matter what `Cargo.toml` and `Cargo.lock` file is being used, we can the metadata from anywhere within a cargo workspace and parse settings. In the context of supporting workspaces, this would allow users to keep `CrateSettings` in the relevant workspace member toml files and have them be joined in an intelligent way to be used for the rest of the `raze` command. Additionally, should we need to locate the cargo files, `workspace_root` should always be the directory where the cargo files live for currently supported use cases (single/root package projects).

Somewhat related to this issue, there's work in https://github.com/UebelAndre/cargo-raze/commit/332ad931f374ea28e3100fe46c92e99dd51b6ae2 which is inspired by conversations in https://github.com/google/cargo-raze/pull/222 but also starts to point out that the `CargoWorkspaceFiles` is not needed. But for now I feel strongly that `override_lockfile` represents something different than `CargoWorkspaceFiles`. Though I agree with your point. I'd like to either delete the need for `override_lockfile` and have everything live in normal Cargo files or delete `CargoWorkspaceFiles` and use Cargo Metadata to gather the necessary information

acmcarther replied:
¯\\\_(ツ)_/¯

acmcarther commented on impl/src/metadata.rs:
```diff
@@ -50,25 +97,132 @@ pub struct CargoWorkspaceFiles {
   cargo_bin_path: PathBuf,
-  use_tempdir: bool,
+  registry_url: Url,
+  index_url: String,
```
Ignored this comment?

UebelAndre replied:
I made it a `Url` I think it's the cleaner implementation

acmcarther commented on impl/src/metadata.rs:
```diff
@@ -50,25 +97,132 @@ pub struct CargoWorkspaceFiles {
+  pub fn new<P: Into<PathBuf>>(
+    cargo_bin_path: P,
+    registry_url: Url,
+    index_url: &str,
```
Ignored this comment?

GregBowyer replied:
Typically from my experience yes, `&String` would suggest that you are giving a reference directly to an owned thing, at which point since I dont get ownership of it in any meaningful way (`Box<String>` / `String` / `&mut String`) I am curious why not have the more primal type.

In the case of avoiding a copy but allowing it if needed I would more expect to see `Cow<'a, str>` as the type/

My two cents, I have not seen much `&String` in the wild

acmcarther replied:
I didn't mean &String, i meant String. Move the value.

The motivation here is basically that we already need an owned string here, but we take an &str because generally `&str` is what you want.

Here though that can result in an extra allocation if the end user no longer needed the string anyway.

UebelAndre replied:
I've come across both cases. I'm not quite sure what all the trade-offs here are, ownership is the only one I've really had experience with. But I think you're right @acmcarther I think it'd be better to have the signature accept `String` to clearer communicate the intent and not mask any allocations.

acmcarther commented on impl/src/metadata.rs:
```diff
@@ -50,25 +97,132 @@ pub struct CargoWorkspaceFiles {
-    let mut url = url::Url::parse(registry_url)?;
+  // Create a symlink file on unix systems
+  #[cfg(target_family = "unix")]
+  fn make_symlink(&self, src: &Path, dest: &Path) -> Result<()> {
```
Ignored this comment?

UebelAndre commented on impl/src/rendering/bazel.rs:
```diff
@@ -301,17 +338,26 @@ impl BuildRenderer for BazelRenderer {
+    let crates_bzl_pkg_file = path_prefix.as_path().join("BUILD.bazel");
+    if !file_outputs
+      .iter()
+      .any(|output| output.path.eq(&crates_bzl_pkg_file))
```
Can-do

acmcarther commented on impl/src/metadata.rs:
```diff
@@ -19,16 +19,70 @@ use std::{
+// These attriburtes are the source of truth for the Metadata used by Cargo raze.
+// Making them public allows for dependents to always ensure they're using the 
+// same versions `cargo-raze` is.
+pub use cargo_metadata::{Metadata, Package, PackageId};
```
Not sure why the comment disappeared, but RE your last response.

Lets not reexport this. If an end users needs this, we can decide to reexport it at that time once a specific case has been made.

UebelAndre replied:
This has been updated and is no longer public.

acmcarther commented on impl/src/testing.rs:
```diff
@@ -215,5 +266,9 @@ pub fn mock_crate_index(crates: &HashMap<String, String>) -> TempDir {
   }
 
-  index_url_mock_dir
+  if mock_dir.is_none() {
```
It seems this function always returns Some(x). Update it to not return Option anymore?

GregBowyer replied:
maybe 

```rust
mock_dir.or_else(|| Some(index_url_mock_dir))
```

UebelAndre replied:
No, it does return `None` some in cases
```
test planning::tests::test_binary_dependencies_vendored_genmode ... FAILED
test planning::tests::test_binary_dependencies_remote_genmode ... FAILED
```
I added a panic on that line and these tests tripped it.

UebelAndre commented:
@acmcarther Don't know why comments are being hidden (maybe the rebasing?) but I think I've gone over everything now.

GregBowyer commented on impl/src/bin/cargo-raze.rs:
```diff
@@ -76,16 +80,21 @@ fn main() -> Result<()> {
-    Some(ref p) => Box::new(CargoMetadataFetcher::new(p, /*use_tempdir: */ true)),
+  // Fetch metadata
+  let metadata_fetcher: Box<dyn MetadataFetcher> = match options.flag_cargo_bin_path {
+    Some(ref p) => Box::new(CargoMetadataFetcher::new(
```
Nit: I might be a bit famous for making variable names smaller (scope-declares-verbosity) but even by my standards I think this can be something like `path`

UebelAndre replied:
Sorry, can you elaborate?

GregBowyer replied:
So I _personally_ follow the practise of making my variable names as descriptive as possible without going to the point of utter verbosity.

Basically the closer a variable is in scope to where it will be used the shorter I feel I can get away with.

For example `x` / `y` is ok for say

```rust
some_iter.map(|x| x.thing)
```

whereas say for your code since `p` is a little further away from its use I might have mental forgotten what `p` relates to by the time it gets used, so in that case I would go with `path`

If it was, yet further away I might go longer and use `metadata_path` (and so on and so forth).

I will make exceptions if the context while short is just utterly obvious or ubiquitous e.g. https://github.com/simh/simh/blob/master/VAX/vax_cpu.c#L631 (I work on this :/) where sure `R` is very short but if you read any documentation on a vax its _the name_ for registers.

What I aim to avoid is either this https://github.com/simh/simh/blob/master/VAX/vax_cpu.c#L3792 (Great `i` is used everywhere and means a million little contextual things) or the other way https://chromium.googlesource.com/native_client/nacl-gcc/+/ng/master/libgo/go/net/fd_windows.go#36 (`hasLoadSetFileCompletionNotificationModes`)

In this context I suggest that `p` is a little too short as I found I had forgotten what `p` was by the time it was used.

UebelAndre replied:
That was a good read. Thanks 😄 

I've updated the variable name.

acmcarther commented on impl/src/bin/cargo-raze.rs:
```diff
@@ -181,52 +192,34 @@ fn write_to_file(path: &Path, contents: &str, verbose: bool) -> Result<()> {
+  bazel_root: &Path,
+  settings: &'settings RazeSettings,
+) -> RemoteGenModeInputs<'settings> {
+  if settings.genmode == GenMode::Remote {
```
Nit: This is cleaner structured as an early return:

```rust
if settings.genmode != GenMode::Remote {
    return RemoteGenModeInputs {
      override_lockfile: None,
      binary_deps: None,
    }
}
// the rest

UebelAndre replied:
Updated

acmcarther reviewed (commented):
functions are too long man

too long

UebelAndre commented:
https://github.com/google/cargo-raze/pull/276#discussion_r521759396
Ah! I see, I thought you were saying I had duplicate code within the same function.

acmcarther commented:
Humble request to stop rebasing the commits unless you have a good reason. It's wrecking the review history.

I try to click https://github.com/google/cargo-raze/pull/276/files/558e3cb7b2259a369880d26642269f376801b73c#diff-c8fe41d4c65c6cc6fc8160d444cb04842ac218c135768f6e4309b18ec27ee784 because you asked me to provide an example, and the link is broken

GregBowyer commented on impl/src/planning/crate_catalog.rs:
```diff
@@ -198,39 +184,35 @@ impl CrateCatalog {
       .iter()
-      .cloned()
-      .collect::<HashSet<_>>();
+      .filter_map(|node| {
```
Could just be 

```rust
filter(|node| metadata.workspace_members.contains(&node.id))
```

no ?

GregBowyer commented on impl/src/planning/subplanners.rs:
```diff
@@ -143,40 +134,44 @@ impl<'planner> WorkspaceSubplanner<'planner> {
+        .metadata
+        .workspace_members
+        .iter()
+        .filter_map(|pkg_id| {
```
I might be missing something but why not

```rust
self
.metadata
.metadata // <-- Also why this stutter?
.packages
.iter()
.find(|pkg| pkg.id == &pkg_id) // <-- Also why the javaesq `.eq(`
.map(|pkg| get_workspace_member_path(&pkg.manifest_path, &self.metadata.metadata.workspace_root))
```

UebelAndre replied:
No, this needs to iterate workspace members specifically.

The 'stutter' is is the name `metadata` being used in two different contexts. One is the glob of metadata output by the metadata fetcher and the inner member is the `cargo_metadata::Metadata`.

GregBowyer commented on impl/src/planning/subplanners.rs:
```diff
@@ -366,13 +314,40 @@ impl<'planner> CrateSubplanner<'planner> {
+          .metadata
+          .packages
+          .iter()
+          .find(|pkg| pkg.id.eq(pkg_id))
```
Same as elsewhere is there a reason this is not `==` am I missing something or being dumb?

UebelAndre replied:
I found this to be clearer than de-referencing references to allow for `==` but that may be personal experience. I've since updated these.

GregBowyer commented on impl/src/rendering/bazel.rs:
```diff
@@ -220,18 +219,32 @@ impl BuildRenderer for BazelRenderer {
+        })
+        .map(|ctx| ctx.clone())
+        .collect();
+      let rendered_alias_build_file = self
```
Nit: I would put a line break between the two bindings to avoid everything becoming a wall of text over time.

UebelAndre replied:
Done

GregBowyer commented on impl/src/testing.rs:
```diff
@@ -45,6 +46,48 @@ dependencies = [
+"#
+}
+
+pub fn advanced_lock() -> &'static str {
```
I would use `indoc` to make the function / const indented. I would also consider a const here

UebelAndre replied:
This was done to be consistent with what was there before (`basic_toml` and `basic_lock`) but I can update it.

UebelAndre replied:
How would I apply `const` here? I thought that was implicit

UebelAndre replied:
I've updated these now.

GregBowyer replied:
```rust
pub const fn advanced_lock()
```

or 

```rust
const ADVANCED_LOCK: &str = indoc! { r#" str str str "#}
```

UebelAndre commented:
> Humble request to stop rebasing the commits unless you have a good reason. It's wrecking the review history.
> 
> I try to click https://github.com/google/cargo-raze/pull/276/files/558e3cb7b2259a369880d26642269f376801b73c#diff-c8fe41d4c65c6cc6fc8160d444cb04842ac218c135768f6e4309b18ec27ee784 because you asked me to provide an example, and the link is broken

Ah, sorry, I'll add new commits from here on without rebasing 😅

UebelAndre commented:
> [#276 (comment)](https://github.com/google/cargo-raze/pull/276#discussion_r521759396)
> Ah! I see, I thought you were saying I had duplicate code within the same function.

This is updated

UebelAndre commented:
https://github.com/google/cargo-raze/pull/276#discussion_r521757623
So would documentation suffice or what should I do?

UebelAndre commented:
@acmcarther @GregBowyer Ready for your eyes again

UebelAndre commented:
@acmcarther Just wanted to thank you for all the attention you gave this PR on Wednesday. Super appreciated! 🙏 

I was wondering if we'd be able to do something similar today. There are more changes I'd like to make but are all done on top of the changes here. I'm very happy with the general state of the project after the work here which were inspired by all of the thoughts you've shared about the design goals of `cargo-raze`. I truly believe that this is my final large PR since I can't think of any major features missing after this change. I would very much like to get this through to unblock other work. As always, looking forward to hearing from you 😄

UebelAndre commented:
@acmcarther I did one more small update. This is ready for review 😅

UebelAndre commented:
@acmcarther Hey, If you don't have the bandwidth for reviewing this, would you be able to recommend another reviewer? I totally understand that free time can be hard to come by but I'd really like to get this functionality in 😅 🙏

acmcarther commented on impl/src/bin/cargo-raze.rs:
```diff
@@ -96,83 +105,85 @@ fn main() -> Result<()> {
     lock_path_opt,
   };
 
+  // Determine the path to the Bazel workspace root or fallback to the current working directory
```
Purely aesthetic thing, but it looks like your inline comments here are meant to describe blocks of code, but there are newlines sprinkled throughout which makes the association not obvious.

If 
```rust
 // Determine the path to the Bazel workspace root or fallback to the current working directory
```
is meant to describe L109 - L117, I think you should remove the newlines:

```rust
  let cargo_raze_working_dir = find_bazel_workspace_root().unwrap_or(std::env::current_dir()?);
  let remote_genmode_inputs = gather_remote_genmode_inputs(&cargo_raze_working_dir, &settings);
  let metadata = metadata_fetcher.fetch_metadata(
    &files,
    remote_genmode_inputs.binary_deps,
    remote_genmode_inputs.override_lockfile,
  )?;
```

The same applies to the `// Do Planning` comment, as well as the `// Render BUILD files` comment, as well as the `// Write BUILD files` comment.

UebelAndre replied:
Do you think that comment is still necessary with the renamed comment? I really felt the `// Section ---` style clearly communicated that the `main` method was sectional and made those comments distinct from others describing functionality

acmcarther replied:
The essential aspect of my comment is that "// some comment" doesn't obviously apply to multiple lines of code once you insert newlines between the lines of code.

The "// whatever ---" seems completely arbitrary. Arbitrary conventions don't really communicate intent very well.

UebelAndre replied:
Ah, I see. So what's the action item from here on?

acmcarther replied:
I'm saying don't do the "---" thing. When a comment applies to multiple lines, you have two options.

1. Keep the lines of code together. Don't add whitespace
2. Move the lines into a function.

acmcarther commented on impl/src/bin/cargo-raze.rs:
```diff
@@ -96,83 +105,85 @@ fn main() -> Result<()> {
   Ok(())
 }
 
+/** Helper function for writing rendered files to disk */
```
This is a pretty extreme nit, but I'm making the comment because it applies in multiple places.

"Helper function" or "helper struct" in the function or type documentation add very little. Trimming them has more impact:

```
/** Writes rendered files to filesystem. */
```

```
/** Represents the inputs specific to the Remote genmode. */
```

```
/** Gathers inputs for the `genmode = "Remote"` builds. */
```

acmcarther commented on impl/src/planning.rs:
```diff
@@ -47,251 +45,157 @@ pub trait BuildPlanner {
-  binary_deps_tempdir: Result<TempDir, io::Error>,
+pub struct BuildPlannerImpl {
+  pub metadata: RazeMetadata,
+  pub settings: RazeSettings,
```
Why is this stuff now pub?

acmcarther replied:
(i'm making this comment not because I have deep wisdom about this, but because it looks superficially like you might be breaking the abstraction to take an implementation shortcut)

UebelAndre replied:
It didn't need to be. Updated

acmcarther replied:
This still looks pub as of most recent commit.

UebelAndre replied:
Sorry 😢 I did forget to add that file.

acmcarther reviewed (changes requested):
I think you missed my (and GregBowyer's) comments which are included in unexpanded files. At least, that's the case with `metadata.rs`. 

Did you move functions around in metadata.rs? The diff makes it look like you basically deleted everything and rewrote it.

acmcarther commented:
Sorry for the delay in the review by the way. I recognize the huge effort you've put in both in this PR and in previous PRs.

UebelAndre commented:
@acmcarther Ok, I think I've found all the comments. Still learning how to use GitHub. I wonder if there's a user setting to always show comments in the "Conversation" tab. I assumed everything showed up there and that I could treat each comment like a TODO list and resolve conversations as appropriate. I'm definitely not ignoring comments. Maybe just going through the feedback a bit too quickly 😅 

This is ready for another review 🙏 I think it's getting closer!! 😄

acmcarther commented on impl/src/settings.rs:
```diff
@@ -369,6 +387,51 @@ fn validate_settings(settings: &mut RazeSettings) -> Result<(), RazeError> {
 
+/** Parses raze settings from the contents of a `Cargo.toml` file */
+fn parse_raze_settings(toml_contents: CargoToml) -> Result<RazeSettings, RazeError> {
+  // Workspace takes precidence
```
Nit s/precidence/precedence

acmcarther commented on impl/src/settings.rs:
```diff
@@ -369,6 +387,51 @@ fn validate_settings(settings: &mut RazeSettings) -> Result<(), RazeError> {
+  {
+    if toml_contents.raze.is_some() {
+      eprintln!(
+        "WARNING: Both [raze] and [pakcage.metadata.raze] are set. Using [package.metadata.raze] \
```
Nit: s/pakcage/package

acmcarther commented on impl/src/settings.rs:
```diff
@@ -369,6 +387,51 @@ fn validate_settings(settings: &mut RazeSettings) -> Result<(), RazeError> {
+  // Workspace takes precidence
+  if let Some(raze) = toml_contents
+    .workspace
+    .and_then(|pkg| pkg.metadata.and_then(|data| data.raze))
```
This appears to add a third possible configuration location:

- [raze]
- [package.metadata.raze]
- [workspace.metadata.raze]

This was not mentioned in the PR description. Why is this being added?

UebelAndre replied:
This PR was opened before https://github.com/google/cargo-raze/pull/274 was merged and I had to make this change after a rebase. I just slipped my mind to add it to the PR description. This is being added because it makes sense to me when authoring projects using cargo workspaces. Associating raze settings with a package gets confusing when there's no root package. To solve for this case, we should use `[workspace.metadata]`

UebelAndre replied:
I also updated the PR description.

acmcarther commented on impl/src/planning/subplanners.rs:
```diff
@@ -366,13 +312,40 @@ impl<'planner> CrateSubplanner<'planner> {
+          .metadata
+          .packages
+          .iter()
+          .find(|pkg| pkg.id == *pkg_id)
```
Nit: This expression is too long.

```rust
    let workspace_package = self
        .crate_catalog
        .metadata
        .packages
        .iter()
        .find(|pkg| pkg.id == *pkg_id);
    if let Some(workspace_package) = workspace_package {
      // etc
    } else {
      // etc
    }
  ```

UebelAndre replied:
Updated

UebelAndre commented:
> I think you missed my (and GregBowyer's) comments which are included in unexpanded files. At least, that's the case with `metadata.rs`.
> 
> Did you move functions around in metadata.rs? The diff makes it look like you basically deleted everything and rewrote it.

I don't remember doing a large shuffle. But I did just restore the placement of one thing and split some functionality from `make_temp_workspace` into it's own function. Trying to keep things minimal now

acmcarther commented on impl/src/planning/subplanners.rs:
```diff
@@ -143,40 +134,44 @@ impl<'planner> WorkspaceSubplanner<'planner> {
+            .metadata
+            .packages
+            .iter()
+            .find(|pkg| pkg.id == *pkg_id)
```
Nit: Made this comment elsewhere, duplicating here.

Can we avoid these extremely long conditional clauses?

```rust
let workspace_member = self
            .metadata
            .metadata
            .packages
            .iter()
            .find(|pkg| pkg.id == *pkg_id);
if let Some(workspace_member) = workspace_member {
```


```

UebelAndre replied:
I'll keep this in mind for future changes.

acmcarther commented on impl/src/planning/subplanners.rs:
```diff
@@ -101,39 +101,30 @@ pub struct WorkspaceSubplanner<'planner> {
 impl<'planner> WorkspaceSubplanner<'planner> {
   /** Produces a planned build using internal state. */
   pub fn produce_planned_build(&self) -> Result<PlannedBuild> {
+    // Check for errors ---
```
This "---" is not a convention of comment writing, as far as I know. I prefer regular comments and avoiding interstitial newlines to communicate that a comment refers to a whole block of code.

UebelAndre replied:
Updated

acmcarther reviewed (changes requested):
I have the sneaking suspicion that this is going to to be a breaking change, but the test suite isn't good enough to catch it, and I'm no longer familiar enough with the code to identify it in review.

Oh well I guess. I think this is the last pass of comments.

UebelAndre commented:
@acmcarther Updated. Back to you 🙏

UebelAndre commented:
> I have the sneaking suspicion that this is going to to be a breaking change, but the test suite isn't good enough to catch it, and I'm no longer familiar enough with the code to identify it in review.
> 
> Oh well I guess. I think this is the last pass of comments.

This may be a "breaking" change for more obscure projects. Notable the generation of the aliases is controlled by the `workspace_member_dir` setting and is always relative to the location of `Cargo.toml` files. That said though, I think with that setting, users should be able to get new generations with these changes to match previous outputs.

UebelAndre commented:
Fixed a typo which should fix CI.

UebelAndre commented:
https://github.com/google/cargo-raze/pull/277 is also green with no changes other than a rebase

UebelAndre commented:
@acmcarther Hey, are you around to take another quick look? 😅

acmcarther commented on impl/src/bin/cargo-raze.rs:
```diff
@@ -76,16 +80,21 @@ fn main() -> Result<()> {
     None => Box::new(CargoMetadataFetcher::default()),
   };
-  let mut planner = BuildPlannerImpl::new(&mut *metadata_fetcher);
 
```
Still a bunch of whitespace here between L89 "fetch metadata" and L118 "do planning".

UebelAndre replied:
Sorry, I thought removing the comment form the earlier conversation meant this was not necessary. Updated

acmcarther commented on impl/src/bin/cargo-raze.rs:
```diff
@@ -96,83 +105,84 @@ fn main() -> Result<()> {
-  let planned_build = planner.plan_build(&settings, &prefix_path, files, platform_details)?;
+  // Render BUILD files
   let mut bazel_renderer = BazelRenderer::new();
 
```
Whitespace?

UebelAndre replied:
Updated

UebelAndre commented:
@acmcarther Removed the whitespace. Ready for another pass 🙏

acmcarther commented:
Will wait for green from CI

UebelAndre commented:
@acmcarther Just finished 😄

UebelAndre commented:
@acmcarther thank you so much for all your support!!! 🙏

dae commented:
@UebelAndre would it make sense for workspace_member_dir to default to "." for compatibility? After updating I found I was getting cargo/cargo/BUILD.bazel instead of cargo/BUILD.bazel, and had to dig around to find out how to change it.

UebelAndre commented:
@dae I can do that but I originally worried about setting it to `"."` because I was afraid of overriding valuable build files.

Also, what do you think of the name `workspace_member_dir`? It's really just the Bazel package aliases should be rendered into. Is there a  better name for it?

dae commented:
Would `build_file_output_dir` work, or does it have other uses I have missed?

UebelAndre commented:
@dae Also, is your project public? I'd love to know more about how raze is used. From what I've gathered, there's two kinds of usage:
1. Cargo projects that are being ported to bazel.
2. Bazel projects that want to introduce dependencies on crates.io crates. 

I've been focusing on the first case for some time which is partly why `workspace_member_dir` ended up in the state that it is. I'd love to get to a point where either users don't have to define bogus `Cargo.toml` files to use this tool. Then for the second point, I'd want to update `rules_rust` to be able to act as the authority and interract with crate registries from within Bazel entirely. No `Cargo.toml` files required. Users should be able to author their targets to define all the dependencies they require and the rules should efficiently handle the rest. I think there's a very real path forward to that goal with this tool and maybe some updates to `rules_gazelle`. 

Anywho, I'd love to see your project structure and to know which camp you fall into (camp 1 or camp 2) to give me better context for future changes.

UebelAndre commented:
> Would `build_file_output_dir` work, or does it have other uses I have missed?

I feel like that's similar to `workspace_member_dir` in that it doesn't quite communicate the intent. I was thinking something like `package_aliases_dir`. Your thoughts there?

dae commented:
`aliases_build_file_dir`? `top_build_file_dir`? 

I run an open source project that was recently converted to Bazel, though it does not currently use workspaces: https://github.com/ankitects/anki

dae commented:
(the point being that when the BUILD file moved, I was hunting for something related to the build file that controlled where it was placed)

UebelAndre commented:
> (the point being that when the BUILD file moved, I was hunting for something related to the build file that controlled where it was placed)

Would I be correct in saying your primary frustration was in the lack of documentation? This I can understand and I'm going to be writing documentation ASAP if that's the case or not.  I've created https://github.com/google/cargo-raze/pull/306 to start addressing this feedback. We should continue the conversation there.

Also, thanks for showing me your project! 🙏
